### PR TITLE
Faster addition using Intel intrinsics.

### PR DESCRIPTION
**Pull request by jzmaddock**
Replaces: https://github.com/boostorg/multiprecision/pull/233.

### CODE REVIEW

**Comment by jzmaddock** [comment]
Just discovered (while looking at subtraction) that there's a subtle issue with the benchmarks we were using that meant we were testing non-random values.  I need to re-evaluate the benchmarks just as a double check...

**Comment by jzmaddock** [comment]
Benchmarks in PR's prior to this one were generating test values which were mostly zero (so not carry!).

Updated MSVC test results:

```
BM_add_old<cpp_int>/1024            73160 ns        71150 ns        11200
BM_add_old<cpp_int>/2048           115318 ns       114397 ns         5600
BM_add_old<cpp_int>/4096           171206 ns       167426 ns         3733
BM_add_old<cpp_int>/8192           310536 ns       307603 ns         2489
BM_add_old<cpp_int>/16384          726716 ns       739397 ns         1120
BM_add_old<cpp_int>/32768         1161796 ns      1143973 ns          560
BM_add_old<cpp_int>/65536         1950260 ns      1968834 ns          373
BM_add_new<cpp_int>/1024            30711 ns        31145 ns        23579
BM_add_new<cpp_int>/2048            47348 ns        48652 ns        14452
BM_add_new<cpp_int>/4096            80554 ns        81961 ns         8960
BM_add_new<cpp_int>/8192           151841 ns       149972 ns         4480
BM_add_new<cpp_int>/16384          326254 ns       329641 ns         2133
BM_add_new<cpp_int>/32768          769317 ns       784738 ns          896
BM_add_new<cpp_int>/65536         1552230 ns      1569475 ns          448
BM_add_new_64<cpp_int>/1024         22431 ns        21973 ns        32000
BM_add_new_64<cpp_int>/2048         36448 ns        36622 ns        21333
BM_add_new_64<cpp_int>/4096         52126 ns        51563 ns        10000
BM_add_new_64<cpp_int>/8192         90180 ns        89979 ns         7467
BM_add_new_64<cpp_int>/16384       220225 ns       219727 ns         3200
BM_add_new_64<cpp_int>/32768       664631 ns       662667 ns          896
BM_add_new_64<cpp_int>/65536      1439122 ns      1429967 ns          448
```
The BM_add_new_64 tests do 64-bit adds even though the limbs are 32 bit - so there is a slight but probably worthwhile improvement.

Updated benchmark:

```
#include <iostream>
#include <benchmark/benchmark.h>
#include <boost/multiprecision/cpp_int.hpp>
#include <boost/multiprecision/gmp.hpp>
#include <boost/random.hpp>
#include <cmath>

#include <immintrin.h>

using namespace boost::multiprecision;
using namespace boost::random;

namespace boost {
   namespace multiprecision {
      namespace backends {


         template <class CppInt1, class CppInt2, class CppInt3>
         inline BOOST_MP_CXX14_CONSTEXPR void add_unsigned_64(CppInt1& result, const CppInt2& a, const CppInt3& b) BOOST_MP_NOEXCEPT_IF(is_non_throwing_cpp_int<CppInt1>::value)
         {
#ifndef BOOST_MP_NO_CONSTEXPR_DETECTION
            if (BOOST_MP_IS_CONST_EVALUATED(a.size()))
            {
               add_unsigned_constexpr(result, a, b);
            }
            else
#endif
            {
               using ::boost::multiprecision::std_constexpr::swap;

               // Nothing fancy, just let uintmax_t take the strain:
               unsigned m(0), x(0);
               unsigned as = a.size();
               unsigned bs = b.size();
               minmax(as, bs, m, x);
               if (x == 1)
               {
                  bool s = a.sign();
                  result = static_cast<double_limb_type>(*a.limbs()) + static_cast<double_limb_type>(*b.limbs());
                  result.sign(s);
                  return;
               }
               result.resize(x, x);
               typename CppInt2::const_limb_pointer pa = a.limbs();
               typename CppInt3::const_limb_pointer pb = b.limbs();
               typename CppInt1::limb_pointer       pr = result.limbs();
               //typename CppInt1::limb_pointer       pr_end = pr + m;

               if (as < bs)
                  swap(pa, pb);
               // First where a and b overlap:
               unsigned      i = 0;
               unsigned char carry = 0;
               for (; i + 8 <= m; i += 8)
               {
                  carry = _addcarry_u64(carry, *(unsigned long long*)(pa + i + 0), *(unsigned long long*)(pb + i + 0), (unsigned long long*)(pr + i));
                  carry = _addcarry_u64(carry, *(unsigned long long*)(pa + i + 2), *(unsigned long long*)(pb + i + 2), (unsigned long long*)(pr + i + 2));
                  carry = _addcarry_u64(carry, *(unsigned long long*)(pa + i + 4), *(unsigned long long*)(pb + i + 4), (unsigned long long*)(pr + i + 4));
                  carry = _addcarry_u64(carry, *(unsigned long long*)(pa + i + 6), *(unsigned long long*)(pb + i + 6), (unsigned long long*)(pr + i + 6));
               }
               for (; i < m; ++i)
                  carry = ::boost::multiprecision::detail::addcarry_limb(carry, pa[i], pb[i], pr + i);
               for (; i < x && carry; ++i)
                  carry = ::boost::multiprecision::detail::addcarry_limb(carry, pa[i], 0, pr + i);
               if (i == x && carry)
               {
                  // We overflowed, need to add one more limb:
                  result.resize(x + 1, x + 1);
                  if (result.size() > x)
                     result.limbs()[x] = carry;
               }
               else
                  std::copy(pa + i, pa + x, pr + i);
               result.normalize();
               result.sign(a.sign());
            }
         }



      }
   }
}

template <class T>
std::tuple<std::vector<T>, std::vector<T>, std::vector<T> >& get_test_vector(unsigned bits)
{
   static std::map<unsigned, std::tuple<std::vector<T>, std::vector<T>, std::vector<T> > > data;

   std::tuple<std::vector<T>, std::vector<T>, std::vector<T> >& result = data[bits];

   if (std::get<0>(result).size() == 0)
   {
      mt19937                     mt;
      uniform_int_distribution<T> ui(T(1) << (bits - 1), T(1) << bits);

      std::vector<T>& a = std::get<0>(result); 
      std::vector<T>& b = std::get<1>(result);
      std::vector<T>& c = std::get<2>(result);

      for (unsigned i = 0; i < 1000; ++i)
      {
         a.push_back(ui(mt));
         b.push_back(ui(mt));
         if (b.back() > a.back())
            b.back().swap(a.back());
         c.push_back(0);
      }
   }
   return result;
}

template <class T>
std::vector<T>& get_test_vector_a(unsigned bits)
{
   return std::get<0>(get_test_vector<T>(bits));
}
template <class T>
std::vector<T>& get_test_vector_b(unsigned bits)
{
   return std::get<1>(get_test_vector<T>(bits));
}
template <class T>
std::vector<T>& get_test_vector_c(unsigned bits)
{
   return std::get<2>(get_test_vector<T>(bits));
}

template <typename T>
static void BM_add_old(benchmark::State& state)
{
   int                         bits = state.range(0);

   std::vector<T>& a = get_test_vector_a<T>(bits);
   std::vector<T>& b = get_test_vector_b<T>(bits);
   std::vector<T>& c = get_test_vector_c<T>(bits);

   for (auto _ : state)
   {
      for (unsigned i = 0; i < a.size(); ++i)
         add_unsigned_constexpr(c[i].backend(), a[i].backend(), b[i].backend());
   }
}
template <typename T>
static void BM_add_new(benchmark::State& state)
{
   int                         bits = state.range(0);

   std::vector<T>& a = get_test_vector_a<T>(bits);
   std::vector<T>& b = get_test_vector_b<T>(bits);
   std::vector<T>& c = get_test_vector_c<T>(bits);

   for (auto _ : state)
   {
      for (unsigned i = 0; i < a.size(); ++i)
         eval_add(c[i].backend(), a[i].backend(), b[i].backend());
   }
}
template <typename T>
static void BM_add_new_64(benchmark::State& state)
{
   int                         bits = state.range(0);

   std::vector<T>& a = get_test_vector_a<T>(bits);
   std::vector<T>& b = get_test_vector_b<T>(bits);
   std::vector<T>& c = get_test_vector_c<T>(bits);

   for (auto _ : state)
   {
      for (unsigned i = 0; i < a.size(); ++i)
         add_unsigned_64(c[i].backend(), a[i].backend(), b[i].backend());
   }
}


BENCHMARK_TEMPLATE(BM_add_old, cpp_int)->RangeMultiplier(2)->Range(1024, 1 << 16);
BENCHMARK_TEMPLATE(BM_add_new, cpp_int)->RangeMultiplier(2)->Range(1024, 1 << 16);
BENCHMARK_TEMPLATE(BM_add_new_64, cpp_int)->RangeMultiplier(2)->Range(1024, 1 << 16);

BENCHMARK_MAIN();
```

**Comment by madhur4127** [comment]
I believe my benchmark data also had no carry.

Are there gains with other compilers as well?

What about AMD? ARM? ADC is supported by every architecture so this should be the concern of the compiler to generate instructions. I am concerned with benchmarks though on these platforms as they are quite popular ones.

**Comment by jzmaddock** [comment]
>I believe my benchmark data also had no carry.

Right, I think I just copied from your benchmark:

```
      uniform_int_distribution<T> ui(T(1) << (bits - 1) + 1, T(1) << bits);
```

But `T(1) << (bits - 1) + 1` == `T(1) << bits`, and unusually msvc didn't warn about operator precedence so we both missed it.  So all the test values ended up as powers of 2 :(

>Are there gains with other compilers as well?

Haven't tried yet.

>What about AMD? ARM? ADC is supported by every architecture so this should be the concern of the compiler to generate instructions. I am concerned with benchmarks though on these platforms as they are quite popular ones.

AMD is really a clone of the Intel instruction set?  No idea about ARM, right now I'm only really interested in what we can do for Intel/AMD.

**Comment by jzmaddock** [comment]
GCC Mingw64 also looks good:

```
BM_add_old<cpp_int>/1024       83547 ns        83705 ns         8960
BM_add_old<cpp_int>/2048      150024 ns       150663 ns         4978
BM_add_old<cpp_int>/4096      267451 ns       266841 ns         2635
BM_add_old<cpp_int>/8192      511456 ns       515625 ns         1000
BM_add_old<cpp_int>/16384    1005119 ns      1000977 ns          640
BM_add_old<cpp_int>/32768    2037512 ns      1992754 ns          345
BM_add_old<cpp_int>/65536    4075570 ns      4064306 ns          173
BM_add_new<cpp_int>/1024       57140 ns        57199 ns        11200
BM_add_new<cpp_int>/2048      105670 ns       104980 ns         6400
BM_add_new<cpp_int>/4096      168310 ns       168795 ns         4073
BM_add_new<cpp_int>/8192      327354 ns       329641 ns         2133
BM_add_new<cpp_int>/16384     652479 ns       645229 ns          896
BM_add_new<cpp_int>/32768    1519680 ns      1499721 ns          448
BM_add_new<cpp_int>/65536    2604220 ns      2604167 ns          264
```

**Comment by jzmaddock** [comment]
MSVC faster subtract results:

```
BM_subtract_old<cpp_int>/1024       41632 ns        41713 ns        17231
BM_subtract_old<cpp_int>/2048       71214 ns        71498 ns         8960
BM_subtract_old<cpp_int>/4096      128228 ns       128348 ns         5600
BM_subtract_old<cpp_int>/8192      242724 ns       239955 ns         2800
BM_subtract_old<cpp_int>/16384     515532 ns       515625 ns         1000
BM_subtract_old<cpp_int>/32768    1439661 ns      1464844 ns          448
BM_subtract_old<cpp_int>/65536    2791531 ns      2761044 ns          249
BM_subtract_new<cpp_int>/1024       32561 ns        32959 ns        21333
BM_subtract_new<cpp_int>/2048       53102 ns        53125 ns        10000
BM_subtract_new<cpp_int>/4096       87317 ns        85794 ns         7467
BM_subtract_new<cpp_int>/8192      161412 ns       160435 ns         4480
BM_subtract_new<cpp_int>/16384     375951 ns       376607 ns         1867
BM_subtract_new<cpp_int>/32768     975524 ns       983099 ns          747
BM_subtract_new<cpp_int>/65536    1844491 ns      1842752 ns          407
```

**Comment by jzmaddock** [comment]
GCC Minwg results look good too:

```
BM_subtract_old<cpp_int>/1024       47105 ns        47085 ns        14933
BM_subtract_old<cpp_int>/2048       86618 ns        87193 ns         8960
BM_subtract_old<cpp_int>/4096      164086 ns       164958 ns         4073
BM_subtract_old<cpp_int>/8192      323040 ns       322316 ns         2133
BM_subtract_old<cpp_int>/16384     674443 ns       662667 ns          896
BM_subtract_old<cpp_int>/32768    1447775 ns      1443273 ns          498
BM_subtract_old<cpp_int>/65536    2718159 ns      2698293 ns          249
BM_subtract_new<cpp_int>/1024       19173 ns        19252 ns        37333
BM_subtract_new<cpp_int>/2048       32253 ns        32227 ns        21333
BM_subtract_new<cpp_int>/4096       57915 ns        58594 ns        11200
BM_subtract_new<cpp_int>/8192      115059 ns       114397 ns         5600
BM_subtract_new<cpp_int>/16384     296245 ns       291561 ns         2358
BM_subtract_new<cpp_int>/32768     814646 ns       819615 ns          896
BM_subtract_new<cpp_int>/65536    1464284 ns      1464844 ns          448
```

The next problem is that `__has_builtin(__builtin_ia32_addcarryx_u64)` is evaluating to false, even though it is plainly supported on gcc-9 mingw :(

**Comment by madhur4127** [comment]
I am also confused about making this intrinsics work. b2 complaints about undefined reference to addcarry_u64.

BTW why are you checking for ADCX?

**Comment by jzmaddock** [comment]
>BTW why are you checking for ADCX?

GCC defines these functions like this:

```
extern __inline unsigned char
__attribute__((__gnu_inline__, __always_inline__, __artificial__))
_addcarry_u64 (unsigned char __CF, unsigned long long __X,
	       unsigned long long __Y, unsigned long long *__P)
{
  return __builtin_ia32_addcarryx_u64 (__CF, __X, __Y, __P);
}
```

So the intrinsic for `addc` is named `__builtin_ia32_addcarryx_u64` which will suck if in the future they do actually start supporting `adcx`  as they'll have to rename this one :(

**Comment by jzmaddock** [comment]
>I am also confused about making this intrinsics work. b2 complaints about undefined reference to addcarry_u64.

What's the actual error?  Which gcc version etc?  Thanks!

**Comment by madhur4127** [comment]
> GCC defines these functions like this:

I think you're referring to [adxintrin.h](https://github.com/gcc-mirror/gcc/blob/master/gcc/config/i386/adxintrin.h) which also says:

> Never use <adxintrin.h> directly; include <immintrin.h> instead

I don't know but it feels GCC and LLVM don't want people to use this directly :(

I feel like `__x86_64__` and `i386` architecture check should do the job. Because in GCC these intrinsic are in `i386` and this feature is very much available on almost all processors of this century.

**Comment by jzmaddock** [comment]
>Never use <adxintrin.h> directly; include <immintrin.h> instead

I am using <immintrin.h>.

>I feel like __x86_64__ and i386 architecture check should do the job. Because in GCC these intrinsic are in i386 and this feature is very much available on almost all processors of this century.

The addc instruction is, but the _addcarry_u64 intrinsic was only introduced by Intel with the ADX stuff, and by gcc for version 9 and later it seems.

**Comment by jzmaddock** [comment]
@madhur4127 : Is there anything missing from this that your PR has?  Any reason not to merge?  Thanks!

**Comment by jzmaddock** [comment]
Will change immintrin.h to intrin.h on MSVC, though I believe the former is a superset of the latter.

File: include/boost/multiprecision/cpp_int/add_unsigned.hpp [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+#endif
+      carry >>= CppInt1::limb_bits;
+      ++pr, ++pa;
+   }
+   if (carry)
+   {
+      // We overflowed, need to add one more limb:
+      result.resize(x + 1, x + 1);
+      if (result.size() > x)
+         result.limbs()[x] = static_cast<limb_type>(carry);
```
[the remark this answers is not in the document]
Comment (jzmaddock) [reply include/boost/multiprecision/cpp_int/add_unsigned.hpp:80]: Good point, will fix.

File: include/boost/multiprecision/cpp_int/add_unsigned.hpp [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+      for (; i < m; ++i)
+         carry = ::boost::multiprecision::detail::addcarry_limb(carry, pa[i], pb[i], pr + i);
+      for (; i < x && carry; ++i)
+         carry = ::boost::multiprecision::detail::addcarry_limb(carry, pa[i], 0, pr + i);
+      if (i == x && carry)
+      {
+         // We overflowed, need to add one more limb:
+         result.resize(x + 1, x + 1);
+         if (result.size() > x)
+            result.limbs()[x] = carry;
```
[the remark this answers is not in the document]
Comment (jzmaddock) [reply include/boost/multiprecision/cpp_int/add_unsigned.hpp:253]: Will fix.  Good catches, thanks for that!

**Comment by madhur4127** [comment]
I also believe that intrin.h is subset because Intel's intrinsic page points out to use immintrin.h

But the issue is in the documentation that MSVC because it doesn't reflect that.

_A header file, <intrin.h>, is available that declares prototypes for the common intrinsic functions. Manufacturer-specific intrinsics are available in the <immintrin.h> and <ammintrin.h> header files._ from MSVC

**Comment by madhur4127** [comment]
I think this PR fails to compile on GCC.

```
$ gcc -v
Target: x86_64-pc-linux-gnu
gcc version 10.1.0 (GCC)

$ g++ -Wall -Wfatal-errors -Werror -std=c++14 -O3 -I../include bench.cpp -o bench
In file included from ../include/boost/multiprecision/cpp_int/add_unsigned.hpp:10,
                 from ../include/boost/multiprecision/cpp_int/add.hpp:12,
                 from ../include/boost/multiprecision/cpp_int.hpp:2247,
                 from bench.cpp:3:
../include/boost/multiprecision/cpp_int/intel_intrinsics.hpp: In function ‘unsigned char boost::multiprecision::detail::subborrow_limb(unsigned char, boost::multiprecision::limb_type, boost::multiprecision::limb_type, boost::multiprecision::limb_type*)’:
../include/boost/multiprecision/cpp_int/intel_intrinsics.hpp:82:11: error: ‘__builtin_ia32_subborrow_u64’ was not declared in this scope; did you mean ‘__builtin_ia32_sbb_u64’?
   82 |    return __builtin_ia32_subborrow_u64(carry, a, b, reinterpret_cast<cast_type*>(p_result));
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |           __builtin_ia32_sbb_u64
```

This is because GCC uses different SBB.

[GCC's adxintrin.h](https://github.com/gcc-mirror/gcc/blob/master/gcc/config/i386/adxintrin.h)
[LLVM's adxintrin.h](https://github.com/llvm/llvm-project/blob/master/clang/lib/Headers/adxintrin.h)

**Comment by madhur4127** [comment]
```
$ icpc -v                                                         
icpc version 19.1.1.219 (gcc version 10.1.0 compatibility)

$ icpc -Wall -Wfatal-errors -Werror -std=c++14 -O3 -I../include bench.cpp -o bench
In file included from ../include/boost/multiprecision/cpp_int/add_unsigned.hpp(10),
                 from ../include/boost/multiprecision/cpp_int/add.hpp(12),
                 from ../include/boost/multiprecision/cpp_int.hpp(2247),
                 from bench.cpp(3):
../include/boost/multiprecision/cpp_int/intel_intrinsics.hpp(68): error: argument of type "cast_type={unsigned long} *" is incompatible with parameter of type "unsigned long long *"
     return __builtin_ia32_addcarryx_u64(carry, a, b, reinterpret_cast<cast_type*>(p_result));
```

**Comment by jzmaddock** [comment]
>I think this PR fails to compile on GCC.

It's a change in GCC-10 compared to 9.x.  Will fix.

**Comment by jzmaddock** [comment]
The Intel and gcc-10 issues should be fixed now - thanks for the heads up!

I'm still stuck on the clang issue though as I'm unable to reproduce.

**Pull request merged**